opentracing-apache-httpclient/src/main/java/io/opentracing/contrib/apache/ApacheClientSpanDecorator.java

```diff
@@ -0,0 +1,127 @@
+package io.opentracing.contrib.apache;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Logger;
+
+import org.apache.http.Header;
+import org.apache.http.HttpHost;
+import org.apache.http.HttpRequest;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpRequestWrapper;
+import org.apache.http.protocol.HttpContext;
+import org.apache.http.protocol.HttpCoreContext;
+
+import io.opentracing.Span;
+import io.opentracing.tag.Tags;
+
+/**
+ * Decorate span at different stages of request processing. Do not finish span in decorator.
+ *
+ * @author Pavol Loffay
+ */
+public interface ApacheClientSpanDecorator {
+
+    /**
+     * Decorate span before request is fired.
+     *
+     * @param request request
+     * @param httpContext context
+     * @param span span to decorate
+     */
+    void onRequest(HttpRequestWrapper request, HttpContext httpContext, Span span);
+
+    /**
+     * Decorate span after response is received.
+     *
+     * @param response response
+     * @param httpContext context
+     * @param span span to decorate
+     */
+    void onResponse(HttpResponse response, HttpContext httpContext, Span span);
+
+    /**
+     * Decorate span after redirect.
+     *
+     * @param response response
+     * @param httpContext context
+     * @param span span to decorate
+     */
+    void onRedirect(HttpResponse response, HttpContext httpContext, Span span);
+
+    /**
+     *  Decorate span span on error e.g. {@link java.net.UnknownHostException}/
+     *
+     * @param request request
+     * @param httpContext context
+     * @param ex exception
+     * @param span span to decorate
+     */
+    void onError(HttpRequest request, HttpContext httpContext, Exception ex, Span span);
+
+    /**
+     * Decorator which adds standard set of tags and logs.
+     */
+    class StandardTags implements ApacheClientSpanDecorator {
+        private static final Logger log = Logger.getLogger(StandardTags.class.getName());
+
+        @Override
+        public void onRequest(HttpRequestWrapper request, HttpContext httpContext, Span span) {
+            Tags.COMPONENT.set(span, "apache-httpclient");
+            Tags.HTTP_METHOD.set(span, request.getRequestLine().getMethod());
+
+            URI uri = request.getURI();
+            Tags.HTTP_URL.set(span, request.getRequestLine().getUri());
+            Tags.PEER_PORT.set(span, uri.getPort() == -1 ? 80 : uri.getPort());
+            Tags.PEER_HOSTNAME.set(span, uri.getHost());
+        }
+
+        @Override
+        public void onResponse(HttpResponse response, HttpContext httpContext, Span span) {
+            Tags.HTTP_STATUS.set(span, response.getStatusLine().getStatusCode());
+        }
+
+        @Override
+        public void onRedirect(HttpResponse response, HttpContext httpContext, Span span) {
```

Comment:
note you might actually consider a redirect (or rather any network attempt) a different span. notably as the uri socket or even tls might be different. in okhttp brave we use a separate span to represent the application request and one for each network attempt. I haven't done this in brave4 yet as the main thing there was general api style (not nitty gritty stuff like redirects or requests served from cache)